KAN-203, 204-이메일 전송 로직 분리 및 SendGrid 적용

backswan0 opened the pull request:
## #️⃣ CHEERHA Board Number

https://spring-3-jo.atlassian.net/jira/software/projects/KAN/boards/1?selectedIssue=KAN-203
https://spring-3-jo.atlassian.net/jira/software/projects/KAN/boards/1?selectedIssue=KAN-204

## 📝 요약

1. 203: 이메일 전송 로직 분리하기
- 분리한 로직 
   (a) 사용자 및 채용 공고 조회 
   (b) 사용자-채용 공고 매칭 
   (c) 매칭된 데이터 임시 저장 
   (d) 임시 저장된 데이터를 조회하여 이메일 전송 

2. 204: 구글 대신 SendGrid를 적용하여 이메일 알림 기능 구현하기
- 'Too many login attempts' 오류가 발생하여 SendGrid 무료 계정으로 대체 

## 🛠️ PR 유형

- [ ] 배포용 PR : dev 테스트를 완료하였나요?
- [ ] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 논의사항 or 질문

(1) gradle에 있는 firebase는 푸시 알림용 의존성입니다. 삭제하지 말아주세요:)
(2) 이메일 수신자 인증 전이기 때문에 SenderScheduler의 Scheduled 어노테이션을 주석 처리했습니다. 주석 해제 시 이메일이 발송되므로 추가 PR 전까지 해제하지 말아주세요:)
(3) 머지되면 설명 후 SendGrid 사용법 안내 드리겠습니다:D
(4) 머지 이후 이메일 수신자 주소 인증 로직을 더해 추가로 PR할 예정입니다. 

## ✅ PR Checklist

PR이 다음 요구 사항을 충족했는지 확인하기!

- [x] 커밋 메시지 컨벤션 준수
- [x] 변경 사항 테스트 여부 체크(버그 수정/기능 테스트)

## 💥 리뷰어 지목!!!!!

<!--- 리뷰어를 2명 이상 지목해주세요. -->
- [x] 채영
- [ ] 지현
- [x] 리은
- [x] 승찬
- [ ] 주양

backswan0 committed:
:recycle: refactor: 이메일 알림 스케줄러 리팩토링

- Async 어노테이션을 사용하여 논블로킹 처리

- [할 일] submit() 메서드를 사용했을 때와 속도 비교

backswan0 committed:
:recycle: refactor: 이메일 스케줄러 리팩토링 (async 어노테이션 사용)

- 커스터마이징할 필요가 없어서, 즉 공용 스레드 풀을 사용하는 거라서 어노테이션 사용으로 변경

backswan0 committed:
:sparkles: feat: History CSV 파일 generator 클래스 생성

- 사용자 2000명 기준 검색 기록 10000건 생성용 Generator 생성

- 한 사용자가 1번~20번 중 무작위로 검색했다는 조건 추가

backswan0 committed:
:memo: chore: 불필요한 import 삭제

- async 어노테이션을 사용하면서 쓰지 않는 스레드 풀 삭제

backswan0 committed:
:sparkles: feat: 알림 전용 스케줄러 생성

- 스케줄러는 데이터 조회만 담당

- 조회해온 데이터를 임시 저장하는 JPA 레포지토리 생성

backswan0 committed:
:memo: chore: 사용하지 않는 스케줄러 주석 처리 및 히스토리 더미데이터 생성

- 히스토리 더미데이터를 생성하고자 사용자 및 히스토리 개수 수정

- 사용하지 않는 스케줄러 주석 처리

backswan0 committed:
🔨script: yml에 stmp connection pool 추가

- 'Too many login attempts' 문제를 해결하고자 설정 추가

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into V2/refactor/email-non-blocking

backswan0 committed:
:memo: chore: 불필요한 주석 삭제

- 불필요한 import 삭제

- 불필요한 주석 삭제

backswan0 committed:
:sparkles: feat: 알림 발송 전용 클래스 및 스케줄러 생성

- 이메일 발송 전용 클래스 및 실시간 알림 전송 전용 클래스 생성

- 데이터 조회, 매칭, 저장, 저장된 데이터 호출 로직 분리

- [할 일] 지나치게 사용되는 for문 및 foreach문 리팩토링

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into V2/refactor/email-non-blocking

backswan0 committed:
:recycle: refactor: 각 스케줄러 리팩토링

- 의미가 분명하게 클래스 이름 수정

- 이메일은 한 번 전송했으면 값을 true로 변경하여 다시 전송하지 않도록 리팩토링

backswan0 committed:
:recycle: refactor: 매칭된 알림 및 서비스와 레포지토리 리팩토링

- 중복된 Mapping(매칭된 알림) 객체가 다시 저장되지 않도록 로직 추가

- 전송되지 않은 이메일만 조회하도록 메서드 추가

backswan0 committed:
:recycle: refactor: 알림 발송 클래스 및 스케줄러 리팩토링

- 로직마다 주석 추가

- 푸시 알림 구현할 기본 틀 생성

- 알림 객체 중복 생성 예방 및 발송된 알림은 false에서 true 처리 반영

backswan0 committed:
:hammer: script: Firebase 의존성 추가

푸시 알림에 쓰일 FCM 의존성 추가

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into V2/refactor/email-non-blocking

# Conflicts:
#	build.gradle

backswan0 committed:
:fire: delete: 기존에 쓰던 이메일 스케줄러 삭제

- 푸시 알림 반영할 이메일 스케줄러 외에 다른 스케줄러 및 레이어 삭제

- UserDto 파일을 dto 패키지에 이동

backswan0 committed:
Merge remote-tracking branch 'origin/dev' into V2/refactor/email-non-blocking

# Conflicts:
#	src/main/resources/application-dev.yml
#	src/main/resources/application-release.yml

backswan0 committed:
:recycle: refactor: 이메일 알림을 SendGrid로 대체

- 'Too many login attempts' 오류 이슈로 구글에서 SendGrid로 대체

- JavaMailSender 대신 SendGrid 사용

backswan0 committed:
:hammer: script: SendGrid 환경 설정 추가

- yml에 SendGrid 환경 설정 추가

roqkfchqh commented on src/main/java/com/project/cheerha/common/generator/HistoryGenerator.java:
```diff
@@ -0,0 +1,110 @@
+package com.project.cheerha.common.generator;
```
historyGenerator는 풀리퀘에서 빼주세요 ~

backswan0 replied:
넵, 제외하겠습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notice/scheduler/DataFetchScheduler.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Scheduled(cron = "*/30 * * * * *")
+    @Transactional
+    public void fetchData() {
```
무슨 data를 가지고 뭘 하는지 잘 모르겠어요 클래스명이랑 변수명 바꿔주세요

backswan0 replied:
넵 모두 변경하겠습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notice/service/DataFetchService.java:
```diff
@@ -10,17 +10,15 @@
 @Component
 @RequiredArgsConstructor
-public class EmailDataFetchService {
+public class DataFetchService {
```
마찬가지로용

backswan0 replied:
넵, 변경하겠습니다:)

llRosell commented:
코드 확인했습니다 :)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notice/service/EmailSender.java:
```diff
@@ -0,0 +1,127 @@
+
+            mappings.forEach(mapping -> {
+                mapping.markEmailAsSent(); // 발송 완료 상태로 변경
+                mappingRepository.save(mapping); // 변경된 상태 저장
```
트랜잭션이 필요할 거 같은데 private메서드라 지금은 안 걸릴거같은데 고민해보시면 좋을거같습니다

backswan0 replied:
넵, 리팩토링할 때 해당 부분도 고민해보겠습니다:)

roqkfchqh commented on src/main/java/com/project/cheerha/domain/notice/entity/Mapping.java:
```diff
@@ -0,0 +1,70 @@
+        columnNames = {"email", "job_opening_url"}
+    )}
+)
+public class Mapping {
```
뭘 매핑하는건지 잘 모르겠어요 좀 길어지더라도 이름바꿔야될거같아요

backswan0 replied:
넵, 변경하겠습니다:)

backswan0 committed:
:memo: chore: 클래스 및 메서드 이름 수정, 불필요한 클래스 삭제

- HistoryGenerator 삭제

- 클래스가 무엇을 담당하는지 잘 드러나도록 이름 수정

- 어떤 기능인지 잘 보이도록 메서드 이름 수정

tmdcksdl commented:
코드 모두 확인했습니다!
